### PR TITLE
Fix QBI deduction

### LINKED ISSUE #125: Getting the QBI deduction right
@nikhilwoodruff @martinholmer 

We've been trying to make sure our tmd (tax-microdata-benchmarking) file has the best possible information for the qualified business income deduction. From the IRS, we know that total QBI deduction (qbided) was $205.8 billion in 2021 and we've been targeting that. We also know # of returns with qbided and the distribution by AGI range of qbided and number, but have not been targeting them. We've been targeting the qbided total by estimating pass-through W2 wages -- the greater the wages, the greater the possible qbided, up to certain other limits. We've been assuming a simple proportionate relationship between QBI (i.e., the income) and pass-through wages. This does not consider possible more-complex QBI-W2wage relationships, and it does not consider other factors (rules) that can limit qbided.

Using this approach we have been able to solve for a simple pass-through wages proportionate relationship that hits the IRS qbided $205.8 billion with weights that are simply grown from 2015 to 2021 to take into account growth in the number of returns (s006_original in the table below), but that do not reflect reweighting to hit targets.

However, when we reweight (s006 weights in the table below), we have found we cannot hit the $205.8 billion target and so, in the near term, used a simple .33 relationship between QBI and W2 wages (which does not hit the target).

I've looked into this, varying the qbid-related levers that, via Tax-Caculator, affect qbided limits and therefore qbided. The table below summarizes the results. I am racing the clock this morning and can come back to answer questions and edit. The code and table in the next comment should make it clear, for those who want to read that comment. Here's a quick summary; 

The main levers are:

- pass-through W2 wages (PT_binc_w2_wages), which affect a wage-related qbided limit and the wage-and-property-related qbided limit; the higher the value, the higher the possible qbided, unless other limits are hit; in our baseline tmd, we assume they are 33% of QBI
- pass-through property (PT_ubia_property), which affects the wage-and-property-related qbided limit; the higher the value, the higher the possible qbided unless other limits are hit; we assume this is 0 for all filers
- whether the QBI is from a Specified Service Trade or Business (SSTB), in which case the qbided is phased out above certain thresholds, we assume this false for all filers, and so the qbided is not phased out

The table below shows the results of varying these levers and the weights used.

Here are my key takeaways:
-  With baseline assumptions above, and original weights, we can almost hit the IRS target (we have $202 billion) and we know from prior analyses that we had been able to hit it with a different proportion than 33% (and with earlier iterations of the data).
- We are far from it with the "optimized" weights ($154.9 billion)
- We can come close with "optimized" weights if we set w2 wages or ubia property to large values but still can't hit it
- We can go above it with original weights
- This suggests to me (1) our reweighting is a problem, and we've discussed how to address that - namely penalizing large deviations from original weights and need to do that soon, and (2) most taxpayers (I think) probably are not facing significant limits; we probably should raise the assumed w2 wages and ubia property.

More later but must run.

image

Code and table of results are in the next section.

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
Fixes #125

### CODE REVIEW

File: tax_microdata_benchmarking/utils/soi_targets.py
Comment (donboyd5) [inline tax_microdata_benchmarking/utils/soi_targets.py:1]: Clever solution

Comment (nikhilwoodruff) [reply tax_microdata_benchmarking/utils/soi_targets.py:1]: Will be once it works :)

**Comment by martinholmer** [comment]
@nikhilwoodruff,  if in the `imputation_assumptions.py` module, you **change the value of `W2_WAGES_SCALE`  from 0.33 to 0.20**, you get the following results and avold the QBID tax-expenditure test failure:

```
(taxcalc-dev) tax-microdata-benchmarking% pushd tax_microdata_benchmarking ; python create_taxcalc_input_variables.py ; python create_taxcalc_sampling_weights.py ; python create_taxcalc_growth_factors.py ; popd ; pytest . -m taxexp
Creating 2021 PUF+CPS file assuming:
  IMPUTATION_RF_RNG_SEED = 1928374
  IMPUTATION_BETA_RNG_SEED = 37465
  ASSUMED W2_WAGES_SCALE = 0.20000
Skipping CPS previous year income imputation given lack of data...
Importing PolicyEngine US variable metadata...
Uprating PUF from 2015 to 2021...
Pre-processing PUF...
Imputing missing PUF demographics...
Imputing PUF pension contributions...
Constructing hierarchical PUF: 100%|██| 207692/207692 [01:02<00:00, 3322.55it/s]
Creating tc dataset from 'PUF 2021' for year 2021...
Creating tc dataset from 'CPS 2021' for year 2021...
Combining PUF and CPS nonfilers...
>A weights all,puf,cps (#M)= 199.639 161.181 [160.8] 38.458
>A weights all_min,all_max (#)= 1.1 9675.7
>A CTC not in DataFrame
>A W2_wages min,max ($)= 0 52283971
>A weighted puf W2_wages ($B)= 463.053
>A QBID not in DataFrame
Adding Tax-Calculator outputs for 2021...
>B weights all,puf,cps (#M)= 199.639 161.181 [160.8] 38.458
>B weights all_min,all_max (#)= 1.1 9675.7
>B weighted puf CTC ($B)= 250.345
>B W2_wages min,max ($)= 0 52283971
>B weighted puf W2_wages ($B)= 463.053
>B weighted puf QBID ($B)= 273.237 [205.8]
Reweighting...
...reweighting for year 2021
Targeting 514 SOI statistics
Optimising weights: 100%|███████████████████| 2000/2000 [08:43<00:00,  3.82it/s]
...reweighting finished
>C weights all,puf,cps (#M)= 199.312 160.854 [160.8] 38.458
>C weights all_min,all_max (#)= 0.1 16564.3
>C weighted puf CTC ($B)= 176.416
>C W2_wages min,max ($)= 0 52283971
>C weighted puf W2_wages ($B)= 351.592
>C weighted puf QBID ($B)= 206.051 [205.8]    <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Writing PUF+CPS file... [/Users/mrh/work/tax-microdata-benchmarking/tax_microdata_benchmarking/storage/output/tmd_2021.csv]
Writing PUF+CPS file... [/Users/mrh/work/tax-microdata-benchmarking/tax_microdata_benchmarking/storage/output/tmd.csv.gz]
~/work/tax-microdata-benchmarking
============================= test session starts ==============================
platform darwin -- Python 3.11.9, pytest-8.2.0, pluggy-1.5.0
rootdir: /Users/mrh/work/tax-microdata-benchmarking
plugins: dependency-0.6.0, xdist-3.5.0
collected 81 items / 74 deselected / 7 selected                                

tests/test_flat_file.py .......                                          [100%]

================= 7 passed, 74 deselected in 76.06s (0:01:16) ==================
```

cc  @donboyd5

**Comment by nikhilwoodruff** [comment]
Ah, thanks for the catch

File: tests/tc_variable_totals.yaml [hunk 1]
```diff
@@ -33,7 +33,7 @@ e01200: -60216853854.91481
 e01400: 552468020849.5068
 e01500: 1641191913466.6348
 e01700: 1000550811057.0841
-e02000: 927149787725.4258
+e02000: 0.0  # 27149787725.4258
```
Comment (nikhilwoodruff) [inline tests/tc_variable_totals.yaml:36]: Is there a reason for skipping these?

Comment (nikhilwoodruff) [reply tests/tc_variable_totals.yaml:36]: OK sounds good

**Pull request merged**